Create Demo File and Begin Optimizing Code

andrewboldi opened the pull request:
This pull request adds a single file where I will be creating a demo for this 3d-grapher. It is an interactive grapher in which you can enter any valid JS math equation. I have made many optimizations in the code and will start updating the source code itself soon. The demo also includes a slight rotation on its z-axis.

Some of the optimizations I have included are
- Instead of using ```Math.PI / 2```, I simply replaced it with 1.55. 
- multiplication and division by 2 can be done with bit shifting (<< and >>, respectively). 
- Replacing repetitive evaluations such as ```Math.tan(theta)``` with a single calculation, assigning it to a variable, and adding it to all the relevant places.

I do realize that the above replacements are sacrificing code readability, but performance is a top priority for me. Especially since I am motivated to get the surface renderings as clean as possible (no visible squares).

Commit:
add demo file

quantum9Innovation commented:
The demo file idea is great since this can be used as an actual three-dimensional graphing calculator without having to re-code the entire program. The optimizations would also help a ton as to making the code run faster so that it can be re-rendered at a much quicker pace. However, it might be better to use `1.57` instead of `1.55` as an approximation to `Math.PI / 2`. Other than that, all the other optimizations look good!

Commit:
Better decimal precision and more optimizations

Use 1.57 for pi/2 instead of 1.55 🤦‍♂️.

Also, replace the Math.pow(x, 2) with x * x and make numerous other algorithmic optimizations.

Commit:
More optimizations and better lighting

Bit shifting multiplication/division sucks because it ignores floating point numbers. So after removing them and making more optimizations, the graph is almost pristine! The graph renders with much faster and with more detail than before.

Polydynamical commented:
The animation looks very nice and it renders quite quickly too. However, one major issue is the memory management. After saving equations one after another, it significantly slows down. Do you have any ideas on how we might address this?

quantum9Innovation commented:
I think this might be because the old function is still running in the background via the `refresh` function. Right now, the script is constantly calling the `refresh()` function through `window.requestAnimationFrame()` and every time you submit a new equation, a new instance of the `refresh()` function is being called. Instead of using the `refresh` function to animate the graph, it might be better to merge it into the `ieq()` function and call that instead with `window.requestAnimationFrame(ieq)`.

Quantalabs commented:
@quantum9Innovation I think I might have a way to fix this, will work on a fork of @Polydynamical's project, submit a PR there.

Commit:
Optimize some functions

Polydynamical commented:
> @quantum9Innovation I think I might have a way to fix this, will work on a fork of @Polydynamical's project, submit a PR there.

Sounds good, I just made a commit so make sure to pull it before submitting your PR.

Quantalabs commented:
> > @quantum9Innovation I think I might have a way to fix this, will work on a fork of @Polydynamical's project, submit a PR there.
> 
> Sounds good, I just made a commit so make sure to pull it before submitting your PR.

Thanks, will do.

Quantalabs commented:
> What exactly does this do? I just pulled the changes. I'm kind of confused as to why this is here.

GitHub didn't render the comment properly. View the [commit diff](https://github.com/quantum9Innovation/3d-Grapher/pull/5/commits/eed2ee070c4ad47efe4897e85c9a814899d5bdee) to see the comment properly.

Quantalabs committed:
Fix memory management bug and runtime bug

Fix the memory management bug brought to attention by @polydynamical in Quantum9Innovation/3d-grapher#5

Also fix a runtime bug introduced in polydynamical@eed2ee070c4ad47efe4897e85c9a814899d5bdee that prevented users from running program a second time

Fixes:
Fix memory management bug with window.cancelAnimationFrame - https://developer.mozilla.org/en-US/docs/Web/API/Window/cancelAnimationFrame

Fix runtime error through removing what was then line 21, as it was useless, and caused the error.

Quantalabs commented:
Fixed the error, and the another runtime error in Polydynamical/3d-grapher#1

Commit:
Merge pull request #1 from Quantalabs/master

Fix memory management bug and runtime bug

Commit:
Format Code

Commit:
Rename learn.html to demo.html

quantum9Innovation commented on demo.html:
```diff
@@ -0,0 +1,262 @@
+            var i_hat = [1, 0]
+            var j_hat = [1255.77, 0.999]
+
+            var k = 1
```
What exactly do `k` and `v` do?

andrewboldi replied:
I believe `k` is how much to distort the graph in the vertical axis and `v` represents the vertical direction (i.e. should the graph be going up or going down).

quantum9Innovation replied:
Ok, looks good for now

quantum9Innovation commented on demo.html:
```diff
@@ -0,0 +1,262 @@
+                c.fillStyle = "white"
+                c.fillRect(0, 0, w, h)
+
+                graph(f, 64)
```
```suggestion
                graph(f, 32)
```
Using a 32x32 grid will greatly speed up graphing, especially with the animations and everything. If you want, we could add a slider to change the grid size so that it displays better on slower computers.

andrewboldi replied:
Yes, good point. I will add a slider in my next commit.

andrewboldi replied:
However, I do believe 64 squares is a good starting amount since I am easily able to render it with my 1.10 GHZ and 4 GB of RAM laptop.

quantum9Innovation replied:
Great!

quantum9Innovation commented on demo.html:
```diff
@@ -0,0 +1,262 @@
+
+            }
+
+            var brightness = (x, y, z, lTheta, lPhi) => {
```
Just a suggestion, but for the optimizations, removing some of the tedious calculations in these shading functions and replacing them with a simple function would only minimally affect the shading but lead to drastic speed increases.

andrewboldi replied:
Yup! I've added a high/low quality option

quantum9Innovation replied:
Looks good

quantum9Innovation reviewed (commented):
The demo looks great! The only problem is that on some slower computers, the animations don't render quite nicely. I've added a few comments as to new optimizations that we could add, but other than that it's good to merge.

Quantalabs commented on demo.html:
```diff
@@ -0,0 +1,262 @@
+        </div>
+        <canvas style="width: 80%; position: absolute; top: 50%; left: 50%; transform: translate(-50%, -50%);" id="bg">Start streaming on a modern browser!</canvas>
+    </body>
+    <script>
```
Since there's so much JS code, it might be easier to just put this in a separate JS file, like `demo.js` and move the files to `demo/`?
```suggestion    
    <script src="demo.js" type="text/javascript">
```

quantum9Innovation replied:
@Quantalabs can you do this?

Polydynamical replied:
I'll do it

andrewboldi replied:
I'll do it

quantum9Innovation replied:
Okay, ping me so that we can merge when this issue is resolved

andrewboldi replied:
Done

Quantalabs commented on demo.html:
```diff
@@ -0,0 +1,262 @@
+
+                theta += 0.005
+                // 0.005 represent rotation speed
+                animate = window.requestAnimationFrame(refresh)
```
Just for safety:
```suggestion
                const animate = window.requestAnimationFrame(refresh)
```

Quantalabs commented on demo.html:
```diff
@@ -0,0 +1,262 @@
+
+                }
+
+                var nTheta = (theta + 1.570795) % 6.28318
+                var tanNTheta = Math.tan(nTheta)
+                var tanNThetaSquared = tanNTheta * tanNTheta
+                var val2 = Math.sqrt(1 / (tanNThetaSquared + 1))
+                var eval2 = Math.sqrt(-(j_hat[0] * j_hat[0]) + 1)
```
It's better to use:
```suggestion
                let nTheta = (theta + 1.570795) % 6.28318
                let tanNTheta = Math.tan(nTheta)
                let tanNThetaSquared = tanNTheta * tanNTheta
                let val2 = Math.sqrt(1 / (tanNThetaSquared + 1))
                let eval2 = Math.sqrt(-(j_hat[0] * j_hat[0]) + 1)
```

Or if you are not changing them;
```suggestion
                const nTheta = (theta + 1.570795) % 6.28318
                const tanNTheta = Math.tan(nTheta)
                const tanNThetaSquared = tanNTheta * tanNTheta
                const val2 = Math.sqrt(1 / (tanNThetaSquared + 1))
                const eval2 = Math.sqrt(-(j_hat[0] * j_hat[0]) + 1)
```
Since both `let` and `const` add more security.

quantum9Innovation replied:
`const` can't be used here since `theta` changes during the rotation phase. However, `let` would work fine.

Quantalabs commented on demo.html:
```diff
@@ -0,0 +1,262 @@
+        <canvas style="width: 80%; position: absolute; top: 50%; left: 50%; transform: translate(-50%, -50%);" id="bg">Start streaming on a modern browser!</canvas>
+    </body>
+    <script>
+        var equation
```
Wouldn't it be better to do something like:
```suggestion
        let equation
```

quantum9Innovation replied:
Done

Quantalabs reviewed (changes requested):
A lot of the comments have to do with changing `var` to `let`. However, they aren't really required.

Commit:
Add quality options

quantum9Innovation committed:
Safety improvements

Co-authored-by: QLabs <55121845+Quantalabs@users.noreply.github.com>

quantum9Innovation committed:
Use `let` instead of `var`

Co-authored-by: QLabs <55121845+Quantalabs@users.noreply.github.com>

quantum9Innovation commented:
@Quantalabs I am reverting the `const animate = ` commit since that has now stopped users from changing the graph. We can use `let animate = ` instead.

quantum9Innovation committed:
Revert "Safety improvements"

This reverts commit c50f5bffa96d88b8d23bc643714d13eec795e18a.

quantum9Innovation committed:
Revert "Use `let` instead of `var`"

This reverts commit 7f147486d33b527fe52e1ee126a365ebd8cefbad.

quantum9Innovation commented:
After the one remaining issue is fixed, this will be merged onto the `master` branch

Commit:
Create demo.html

Commit:
Create demo.js

Commit:
Split demo.html into demo directory

quantum9Innovation commented:
Okay, looks good; I'll merge it now

Commit:
Update demo.html
